signalfx-java/src/main/java/com/signalfx/signalflow/ChannelMessage.java

```diff
@@ -0,0 +1,345 @@
+package com.signalfx.signalflow;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.fasterxml.jackson.databind.DeserializationFeature;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.signalfx.signalflow.StreamMessage.Kind;
+
+/**
+ * Copyright (C) 2016 SignalFx, Inc. All rights reserved.
+ * 
+ * Base class for stream messages received from a SignalFlow computation.
+ */
+public class ChannelMessage {
+
+    /**
+     * Enumeration of types of channel messages 
+     */
+    public static enum Type {
+
+        STREAM_START(Kind.CONTROL),
+        JOB_START(Kind.CONTROL),
+        JOB_PROGRESS(Kind.CONTROL),
+        CHANNEL_ABORT(Kind.CONTROL),
+        END_OF_CHANNEL(Kind.CONTROL),
+        INFO_MESSAGE(Kind.INFORMATION),
+        METADATA_MESSAGE(Kind.METADATA),
+        DATA_MESSAGE(Kind.DATA),
+        EVENT_MESSAGE(Kind.EVENT),
+        ERROR_MESSAGE(Kind.ERROR);
+
+        private final Kind kind;
+
+        Type(Kind kind) {
+            this.kind = kind;
+        }
+
+        Kind kind() {
+            return kind;
+        }
+    };
+
+    protected static final Logger log = LoggerFactory.getLogger(ChannelMessage.class);
+    protected static final ObjectMapper mapper = new ObjectMapper();
+    static {
+        mapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
+    }
+
+    protected String rawdata;
+    protected ChannelMessage.Type channelMessageType;
+
+    public Type getType() {
+        return this.channelMessageType;
+    }
+
+    public String toString() {
+        return this.rawdata;
+    }
+
+    /**
+     * converts the raw stream message into the proper type of channel message
+     * 
+     * @param streamMessage raw stream message
+     * @return a channel message instance
+     * @throws SignalFlowException if decode fails
+     */
+    public static ChannelMessage decodeStreamMessage(StreamMessage streamMessage)
+            throws SignalFlowException {
+
+        try {
+            ChannelMessage message = null;
+
+            switch (streamMessage.getKind()) {
+
+            case CONTROL:
+                @SuppressWarnings("unchecked")
+                Map<String, Object> dataObject = mapper.readValue(streamMessage.getData(),
+                        Map.class);
+
+                if (Type.STREAM_START.name().equals(dataObject.get("event"))) {
```

Comment:
Hmm, you need to improve your Jackson-fu, because if you deserialize correctly you should be able to get the stream message directly as an instance of `StreamStartMessage` or `JobStartMessage`, etc, without needing this if/elseif/... block. See dynamic typing with Jackson.

For an example in Java, see `ClientMessage.java` in prototype and look at the `@JsonSubTypes` annotation.